Support for problems with scores (PCMS)

brianvar opened the pull request:
![overlay](https://user-images.githubusercontent.com/115399684/201510400-d19254fd-263d-4465-b11e-744fdacb5d01.png)

Basic support for problems with scores and problem overrides on PCMS.

TODO:
- Show difference in queue
- Support for dynamic color of the solved problem with score (red -> green)

Tested on roi-2022-day1

kunyavskiy commented:
Wow. That's cool.

I have several places where I don't live naming.
Especially, ICPCScore vs ICPCBinary. ICPCScore is just not ICPC. 

 Also, I'm not sure if it's a good idea to merge scoreboard/statistics service. I think they will eventually have more differences, than the same. 

I'll add more precise comments later. But please, can you first add your real name and email as author to the proposed commits?

brianvar commented:
@kunyavskiy understood about namings, changed to: `"ICPCBinary" -> "ICPC"`, `"ICPCScore" -> "Score"` https://github.com/icpc/live-v3/pull/81/commits/fb011b472f319d8c5ee9d9575291b413d68ae78f.

I changed the config so future commits will show my data. For now I don't think this is a good idea to change commit author in git history.

What do you mean "I'm not sure if it's a good idea to merge scoreboard/statistics service"? If you are talking about changes in StatisticsService (https://github.com/icpc/live-v3/pull/81/commits/3b145f6024c6f8404587cd582f77b69aed7c4bdc#diff-c9a64c34379dd423fbe6d1bca7e6e9ba7463b71912428a7f8516e7ae02515915) then there was a cast to ICPCProblemResult. I was stuck for several hours why backend is not sending run updates to overlay with score enabled. Later I found this cast and made it dynamic depending on result type. This solved a problem, so for now I think Scoreboard and Statistics connected to each other. Now I made a temporary solution, I think major changes to this we can make in a separate PR.

kunyavskiy commented:
I mean, we probably need separate IOIScoreboardService and IOIStatisticsService. Because current statistics don't make any sense, and the scoreboard becomes overcomplicated.

But I can probably refactor that myself later. 

Is that preparation for Innopolis Open? When it will happen?

brianvar commented:
As I understand it correctly, the Innopolis Open 2023 Elimination 1 is going on right now. I have no connections to the Innopolis Open, I'm just completing tasks for MistonCats and made it in short deadline for today's stream.

I don't think there is any rush now, I may also refactor it in the near future, after I've done other TODO's

irdkwmnsb commented on overlay/src/components/organisms/widgets/Scoreboard.js:
```diff
@@ -92,19 +92,45 @@ const TeamTaskColor = Object.freeze({
+    return <ScoreboardTaskCellWrap background={TeamTaskColor[status]}>
+        {status === TeamTaskStatus.first && <StarIcon/>}
+        {TeamTaskStatus === TeamTaskStatus.unknown ? "?" : score}
+        {status !== TeamTaskStatus.untouched && attempts > 0 && attempts}
```
If a team makes n submissions and gets m score it will be rendered as "mn" without a separator resulting in confusion as to why a score could be more than 100

irdkwmnsb replied:
<img width="386" alt="image" src="https://user-images.githubusercontent.com/8657078/201534896-af8a3131-762c-44d9-bd7d-613323a4a02b.png">
from https://www.youtube.com/watch?v=or5kBuwBxJo&list=PLn4CTt5ibU6c-ykPIbX5OW6TuqxmDL3NK&index=1

brianvar replied:
Fixed in https://github.com/icpc/live-v3/pull/81/commits/ac0e68634fcc9c189456e5f350e638d0b2abacdd

irdkwmnsb commented on overlay/src/components/atoms/ContestCells.js:
```diff
@@ -44,23 +44,53 @@ const VerdictCellWrap = styled(Cell)`
-        {isJudged && result}
+        {data.isFirstToSolve || data.isFirstSolvedRun && <StarIcon/>}
+        {data.percentage !== 0 && !data.isJudged && <VerdictCellProgressBar width={data.percentage * 100 + "%"}/>}
+        {data.isJudged && data.score}
```
When diff is implemented - we should also show it with new score.

irdkwmnsb reviewed (commented):
Nice one.
We were looking into implementing IOI judges for a long time but were overwhelmed with tasks for the WF Dhaka.
I think that the background color of cells in the scoreboard and the queue should represent how well the team solved the contest on a scale from 0 - red, 100 - green.

irdkwmnsb commented:
<img width="504" alt="image" src="https://user-images.githubusercontent.com/8657078/201534948-5e9b3b9a-8909-471c-a93d-7fedba4844a6.png">
Sorting is a bit weird at the moment, but it's an easy fix

irdkwmnsb commented:
From the same livestream - it looks like the service is somehow in emulation - the clock ticks too fast

brianvar commented:
@irdkwmnsb 

> From the same livestream - it looks like the service is somehow in emulation - the clock ticks too fast Is that on purpose?

Yes, we got completed standings.xml and emulated it in 10x speed

> I think that the background color of cells in the scoreboard and the queue should represent how well the team solved the contest on a scale from 0 - red, 100 - green.

Yes, it is written in TODO of the PR.
`Support for dynamic color of the solved problem with score (red -> green)`

Thank you for helping find the bugs, I'll fix them as soon as possible

irdkwmnsb commented on backend/src/main/kotlin/org/icpclive/service/ICPCScoreboardService.kt:
```diff
@@ -126,7 +126,7 @@ abstract class ICPCScoreboardService(optimismLevel: OptimismLevel) {
                 ).also {
                     if (it.score > 0) {
-                        solved++
+                        solved += it.score
```
It's not up to me to judge, but solved is semantically the number of solved tasks – not the overall score.
May break some things somewhere

irdkwmnsb replied:
Also bear in mind that ScoreboardService is considered crutial because depending on it we might show wrong scoreboard and we can't allow for it.

brianvar replied:
I don't see any indications that `totalScore` is the numbers of solved problems. Sorting has been fixed, now it is by the one who scored more points in total. No breakdowns occurred, on the contrary, now the amount in the table is displayed correctly, as well as on https://vk.com/wall-71991592_39972

![image](https://user-images.githubusercontent.com/115399684/201541370-9de91458-38a5-4941-ac6b-d04e5092f627.png)
![image](https://user-images.githubusercontent.com/115399684/201541147-a92588db-2694-4a3b-a029-579b3a7bf7f4.png)

I understand what it's about, but I don't see any reason to create a new variable for total team score and do another sort

brianvar replied:
Even if you look at the use of this variable, it is responsible for sorting, and the meaning has not changed in the conditions. Checks only determine whether the user has solved at least one task - and nothing changes in the situation with Score.

![image](https://user-images.githubusercontent.com/115399684/201541733-46dbd222-91f2-4d61-b199-22da84abb998.png)

irdkwmnsb replied:
Fair enough.
But bear in mind that score can sometimes be a rational number. Like in the Huawei ICPC Challenge.
@kunyavskiy what do you say?

brianvar replied:
@irdkwmnsb rational numbers in scores is not just a problem with `totalScore` and with the fact that `score` is Int, it's also an overlay design problem if score is more than 9. See more on the screenshot in Queue. I made an experimental support for Float's here: https://github.com/icpc/live-v3/pull/81/commits/6ebcc3a1e881d9c2ee0d42216cbf327ecd7492ea.

![image](https://user-images.githubusercontent.com/115399684/201611647-4fd3c19f-a22d-4be2-a29c-d766c6f5762e.png)

Compatibility with integers remained.

irdkwmnsb commented on overlay/src/components/organisms/widgets/Scoreboard.js:
```diff
@@ -104,7 +104,7 @@ const ScoreboardScoreTaskCell = ({ status, score, attempts }) => {
         {status === TeamTaskStatus.first && <StarIcon/>}
         {TeamTaskStatus === TeamTaskStatus.unknown ? "?" : score}
-        {status !== TeamTaskStatus.untouched && attempts > 0 && attempts}
+        {status !== TeamTaskStatus.untouched && attempts > 0 && `-${attempts}`}
```
Now viewers will see 100-3 which might confuse some people. Maybe add a space? Or show attempts inside brackets?
(1000-7 vibes)

brianvar commented:
**Introducing dynamic colors**

![dynamic colors](https://user-images.githubusercontent.com/115399684/201680798-0d39b085-dfde-4c52-b8a4-d302a31b62a9.png)

irdkwmnsb reviewed (approved):
Excellent work!
LGTM!

brianvar commented on cds/src/main/kotlin/org/icpclive/api/ContestInfo.kt:
```diff
@@ -15,6 +15,12 @@ import kotlin.time.Duration
 data class MedalType(val name: String, val count: Int)
 
+@Serializable
+enum class ContestResultType(val type: Int) {
```
Already fixed in https://github.com/icpc/live-v3/pull/81/commits/a03619d478029068742a806056d3c923cc6cf373

brianvar commented on cds/src/main/kotlin/org/icpclive/cds/pcms/PCMSDataSource.kt:
```diff
@@ -34,15 +39,31 @@ class PCMSDataSource(val properties: Properties, creds: Map<String, String>) : F
+
+        var problemsElement = element.child("challenge")
+
+        if(properties.containsKey("problems.url")) {
```
Because we have situations when we can't control standings.xml, it is generated by the organizers of the olymp

brianvar commented on cds/src/main/kotlin/org/icpclive/cds/pcms/PCMSDataSource.kt:
```diff
@@ -25,6 +27,9 @@ class PCMSDataSource(val properties: Properties, creds: Map<String, String>) : F
     }
 
+    private val resultType = ContestResultType.valueOf(
+        properties.getProperty("standings.resultType")?.toString()?.uppercase() ?: "BINARY"
```
Already fixed in https://github.com/icpc/live-v3/commit/a03619d478029068742a806056d3c923cc6cf373

brianvar commented on backend/src/main/kotlin/org/icpclive/service/StatisticsService.kt:
```diff
@@ -19,13 +19,31 @@ class StatisticsService {
-                                results.count { !it.isSolved && it.wrongAttempts > 0 && it.pendingAttempts == 0 },
-                                results.count { !it.isSolved && it.pendingAttempts > 0 },
-                            )
+                            val results = buildList {
```
Okay, I'll try to find a better way

kunyavskiy commented:
I think everything else is minor, and I can it fix it myself after merging.

brianvar commented:
> I think everything else is minor, and I can it fix it myself after merging.

I have some problems with moving to the DifferenceAdapter. It seems to me that the idea of layering adapters on the ContestDataSource isn't good. This is very strange, I call `emit` and pass runs to the main via `complete` PCMSDataSource, however, no events are fired (0 rows in scoreboard). At the same time, the runs themselves from standings.xml are successfully loaded. I think I need to learn more about how it works, I'll do it tomorrow

brianvar commented:
Thank you @irdkwmnsb @kunyavskiy for helping me find the bugs and gave advice how to fix them.

New fields in events.properties for PCMC:
```
standings.resultType = <ICPC | IOI>
standings.minScore = <default: 0>
standings.maxScore = <default: 100>
```

### Tests:

IOI Mode (roi-2022-day1)
![test1](https://user-images.githubusercontent.com/115399684/202414947-3d697dc8-e384-45f6-a07a-c758dca30feb.png)

ICPC Mode (nerc-2021)
![test2](https://user-images.githubusercontent.com/115399684/202415033-c2a6aba3-c003-4fed-bc23-0692693f5e4c.png)

Ready to merge.
